### PR TITLE
Refactor CMake scripts.

### PR DESCRIPTION
**Pull request by trivialfis**
* Demo for using CMake with XGBoost.
* Document for using CMake with XGBoost.
* Use CMake CUDA language.
* Use CMake to handle Doxygen.
* Split up CMakeList.
* Export install target.
* Remove build.sh
* Workaround for gpu_hist test.
* Remove build_config.h.in
* Bump CMake for GPU to 3.12
* Move compilation database related code to tidy.py

TODOs:
- [x] Bring back clang-tidy.
- [x] Enable CLI test in Jenkins.
- [x] Get rabit build on Windows.
- [x] Clean up.
- [x] Fix R installation hack.
- [x] Correct tidy flags.

### CODE REVIEW

**Comment by trivialfis** [comment]
@RAMitchell Thanks for the suggestion, I added your workaround for dmlc registrations.  But the added `inline` keyword for GPU kernel generates a warning.

**Comment by jschueller** [comment]
Hi, it could install the python package in a dir relative to CMAKE_INSTALL_PREFIX (lib/pythonx.ysite-packages on unix and Lib/site-packages on win32).

**Comment by trivialfis** [comment]
@jschueller That would be part of Python setuptools.   I don't think CMake is the right tool to install Python package.

**Comment by jschueller** [comment]
Correct me if I'm wrong but it seems the python bindings is just made of a bunch of .py files, I guess cmake could copy them to the site dir.

**Comment by trivialfis** [comment]
@jschueller setuptools does a lot more than copying scripts.  It helps compile Python scripts into bytecodes, creates a Packages and handles its dependencies.

**Comment by trivialfis** [comment]
@hcho3 Related to #4331.  I bumped CMake to 3.8 but another attempt by @RAMitchell uses 3.9, see #4316.  I'm not sure this PR will work out smoothly as you can see from the tests, my build failed in various ways depending on different platforms, while built perfectly on my machine.  I believe there are lots of different bugs from different platforms ...  So please don't put too much weight on this PR for now,  I'm not sure I can workaround all of them ...

**Comment by hcho3** [comment]
@trivialfis Okay, in that case, should we keep the current CMake version (3.6.0) for now?

**Comment by trivialfis** [comment]
@hcho3 Yep, thanks.  I will upgrade it when it's done.

**Comment by trivialfis** [comment]
List of known issues in CMake CUDA lang:
- Chosen compiler by exporting `CC` and `CXX` doesn't not get propagated to CUDA.
- `CXX_STANDARD` does not get propagated to CUDA.
- Flags like `-Xcompiler xxx` or `-gencode arch,code ...` that contain space does not do well with `target_compile_options`.

**Comment by trivialfis** [comment]
As mentioned by @RAMitchell and @hcho3, I should write a short introduction about rationale about this rewrite.  Here it's. ;-)

# Goal
tl;dr: Enable C APIs with easy to use build system interface:
```
find_package(xgboost REQUIRED)
target_link_libraries(mytarget xgboost::xgboost)
```
The above two lines should get include path and linking done.

# Why it takes a rewrite to accomplish this
There's a trend about [Modern CMake](https://cliutils.gitlab.io/modern-cmake/), where one specify build options target by target.  The extra effort is not only to make the code look nicer, but a fundamental requirement for making a usable shared library.  Since we are building C APIs, user of the interface should not know anything about what libraries xgboost need, what extra flags are required to build a project with xgboost, nor the CXX standard we use.  For instance, users of xgboost should not need to pass `OpenMP` flags to their projects.  That means we need to seal all these stuffs behind the build system.

With `CMake`, all requirements of using a library is abstracted in a installed config file like `xgboost-config.cmake` which is used by `find_package` command.  And the target based style `CMake` is to make sure we only export absolute necessary build specification to our interface.  In xgboost, that's only include directory and the path of linkable object.  Commands like `include_directories`, `link_libraries` and global variables that don't distinguish between *interface* and *internal* should **not** be used any more.

# What problems this won't solve
Symbols visibility.  For a shared library, providing restricted but stable ABI is important for easy distribution.  In some cases one would pass flags like `gcc -fvisibility=hidden` to make all symbols hidden from interface except the ones marked by `__attribute__(visibility("default"))`.  There are other complications in this like whether `dmlc::Error` should be visible to other users.  This PR doesn't address this problem, and not likely in the future since we have to go through `dmlc-core` and `rabit` to make a clear sense about the implications.

**Comment by trivialfis** [comment]
@RAMitchell Could you take a look at the above known issues?  I felt like there are too many things to work around ...

**Comment by hcho3** [comment]
@trivialfis Can we just use a recent version of CMake to avoid known issues?

**Comment by trivialfis** [comment]
@hcho3 Currently I build it with CMake 3.12, which ships with Ubuntu 18.10.  Do you think it's a little too aggressive? @RAMitchell

**Comment by trivialfis** [comment]
Currently it's the last item
>  Failed to link cuda runtime before CMake@3.12. Probably due to our use of object library.

I can't work around.  I don't know how to link CUDA runtime manually.  Another way to do it might be using static library with PIC for libxgboost at first then convert it into shared library later ..  But it's getting more and more weird.

**Comment by hcho3** [comment]
@trivialfis I'm okay with using CMake 3.12 for CUDA code.

**Comment by RAMitchell** [comment]
@trivialfis I am ok with requiring cmake 3.12. I don't have any experience with your particular issue linking the run-time, but in the past I have built the project by combining GPU code into a static library and linking this into the xgboost shared library. This was a workaround for linking issues due to dmlc register, at the time I could not make it work using object libraries.

File: demo/regression/machine.conf [hunk 1]
```diff
@@ -1,30 +1,28 @@
 # General Parameters, see comment for each definition
 # choose the tree booster, can also change to gblinear
 booster = gbtree
-# this is the only difference with classification, use reg:squarederror to do linear classification
+# this is the only difference with classification, use reg:linear to do linear classification
```
Comment (hcho3) [inline demo/regression/machine.conf]: Can you remove `demo/regression/machine.conf` from this pull request? It's not relevant to the CMake refactor.

File: tests/ci_build/Dockerfile.gpu [hunk 2]
```diff
@@ -15,8 +15,8 @@ RUN \
     wget https://repo.continuum.io/miniconda/Miniconda2-4.3.27-Linux-x86_64.sh && \
     bash Miniconda2-4.3.27-Linux-x86_64.sh -b -p /opt/python && \
     # CMake
-    wget -nv -nc https://cmake.org/files/v3.6/cmake-3.6.0-Linux-x86_64.sh --no-check-certificate && \
-    bash cmake-3.6.0-Linux-x86_64.sh --skip-license --prefix=/usr
+    wget -nv -nc https://cmake.org/files/v3.12/cmake-3.12.0-Linux-x86_64.sh --no-check-certificate && \
+    bash cmake-3.12.0-Linux-x86_64.sh --skip-license --prefix=/usr
```
Comment (hcho3) [inline tests/ci_build/Dockerfile.gpu]: Don't forget to update `tests/ci_build/Dockerfile.clang_tidy` as well.

File: demo/regression/machine.conf (hunk 1 above)
Comment (trivialfis) [reply demo/regression/machine.conf]: @hcho3 It's used in a test for CLI.  But I can split it into another PR.

**Comment by trivialfis** [comment]
@hcho3 @RAMitchell Ready for another review.  The build system is quite huge by now,  CMake has its limitation.  If you find anything that is not clear or over complicated in the code, please let me know.  I don't want anyone get stumbled by the build system.

**Comment by trivialfis** [comment]
Added a short intro in `demo/c-api/readme.md` for using CMake with xgboost.

**Comment by trivialfis** [comment]
@RAMitchell Would be be able to try it on Windows with CUDA when time allows?

File: cmake/Doc.cmake [hunk 3]
```diff
@@ -0,0 +1,16 @@
+function (run_doxygen)
+  find_package(Doxygen REQUIRED)
+
+  if (NOT DOXYGEN_DOT_FOUND)
+    message(FATAL_ERROR "Command `dot` not found.  Please install graphviz.")
```
Comment (RAMitchell) [inline cmake/Doc.cmake]: "Please install graphviz." Should this be doxygen?

File: demo/c-api/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+C-APIs
+===
+
+**XGBoost** implements a C API originally designed for various language
+bindings.  For detailed reference, please check xgboost/c_api.h.  Here is a
+demonstration of using the API.
+
+# CMake
```
Comment (RAMitchell) [inline demo/c-api/README.md]: We need tests for this demo using cmake in a future PR. Functionality that is not tested is likely to be broken accidentally.

File: cmake/Doc.cmake (hunk 3 above)
Comment (trivialfis) [reply cmake/Doc.cmake]: @RAMitchell No.  It's the `dot` command used to generate class hierarchy.

File: demo/c-api/README.md (hunk 4 above)
Comment (trivialfis) [reply demo/c-api/README.md]: @hcho3 is changing the ci I think?  To test this we need to create two dummy projects one for each case.

Comment (hcho3) [reply demo/c-api/README.md]: @trivialfis Yes, I'm going to review this PR right now, and then file a PR to refactor Jenkins.

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,359 +1,205 @@
-cmake_minimum_required (VERSION 3.2)
-project(xgboost)
+cmake_minimum_required(VERSION 3.2)
+project(xgboost LANGUAGES CXX C VERSION 0.82)
```
Comment (hcho3) [inline CMakeLists.txt]: Any reason why we put the version number here?

File: cmake/Utils.cmake [hunk 6]
```diff
@@ -29,25 +28,37 @@ function(msvc_use_static_runtime)
...
       endforeach()
+      set(variables
+          CMAKE_CUDA_FLAGS_DEBUG
+          CMAKE_CUDA_FLAGS_MINSIZEREL
+          CMAKE_CUDA_FLAGS_RELEASE
+          CMAKE_CUDA_FLAGS_RELWITHDEBINFO
+      )
+      foreach(variable ${variables})
+          if(${variable} MATCHES "-MD")
+              string(REGEX REPLACE "-MD" "-MT" ${variable} "${${variable}}")
```
Comment (hcho3) [inline cmake/Utils.cmake]: What is this section for? Doesn't the previous section already replace `/MD` with `/MT` for MSVC?

Comment (hcho3) [reply cmake/Utils.cmake]: @RAMitchell Does MSVC use both `/MD` and `-MD` flags?

File: CMakeLists.txt (hunk 5 above)
Comment (trivialfis) [reply CMakeLists.txt]: @hcho3 It will be exported to the CMake install file, used by client CMake script `find_package(xgboost 0.82)`

File: cmake/Utils.cmake (hunk 6 above)
Comment (trivialfis) [reply cmake/Utils.cmake]: @RAMitchell My guess is nvcc uses Unix like `-` syntax on Windows instead of `/`?  So one section for CXX and another for CUDA.  Notice the names of processed variables. `CXX` vs `CUDA`.

File: CMakeLists.txt (hunk 5 above)
Comment (hcho3) [reply CMakeLists.txt]: @trivialfis Would a user want to install multiple versions of XGBoost? Otherwise, the version of XGBoost is determined by the git commit of cloned XGBoost repository.

File: cmake/Utils.cmake (hunk 6 above)
Comment (hcho3) [reply cmake/Utils.cmake]: Got it.

File: CMakeLists.txt (hunk 5 above)
Comment (trivialfis) [reply CMakeLists.txt]: @hcho3 
> Would a user want to install multiple versions of XGBoost

I don't know.  But it seems weird that there's no way to determine the version of a package?  When you try to find your dependency, making sure it's greater than some version seems to be a common practice right?  Like using `find_package(Boost 1.64)` to make sure boost is **at least** 1.64 ...

It's possible to call `git log` then process the strings to get the correct version.  But let leave it as the last resort.

Comment (hcho3) [reply CMakeLists.txt]: @trivialfis Ah so the user can specify the minimum required version of XGBoost. Got it.

File: cmake/Utils.cmake [hunk 7]
```diff
@@ -69,19 +80,23 @@ function(format_gencode_flags flags out)
...
+  set(${out} "${${out}}--generate-code=arch=compute_${ver},code=compute_${ver};")
+
   set(${out} "${${out}}" PARENT_SCOPE)
 endfunction(format_gencode_flags flags)
 
 # Assembles the R-package files in build_dir;
 # if necessary, installs the main R package dependencies;
 # runs R CMD INSTALL.
 function(setup_rpackage_install_target rlib_target build_dir)
+  # backup cmake_install.cmake
```
Comment (hcho3) [inline cmake/Utils.cmake]: What is this for?

Comment (trivialfis) [reply cmake/Utils.cmake]: @hcho3 The next line removes whole R-package build directory, resulting in a missing `cmake_install.cmake` file.  That will cause installation failure.

File: src/tree/updater_gpu_hist.cu [hunk 8]
```diff
@@ -47,8 +49,9 @@ struct GPUHistMakerTrainParam
                   "-1 to use all rows assignted to a GPU, and 0 to auto-deduce");
   }
 };
-
+#if !defined(GTEST_TEST)
```
Comment (hcho3) [inline src/tree/updater_gpu_hist.cu]: So we are excluding parameter registration when `GTEST_TEST` is defined? What benefit does it bring? Is this related to Google Test?

Comment (trivialfis) [reply src/tree/updater_gpu_hist.cu]: @hcho3 No benefit.  Just to workaround the gpu hist test without overriding the file's git history.  See #4322

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  CXX_STANDARD_REQUIRED ON)
+target_compile_definitions(objxgboost
+  PRIVATE
+  -DDMLC_LOG_CUSTOMIZE=1  # enable custom logging
+  $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:_MWAITXINTRIN_H_INCLUDED>
+  ${XGBOOST_DEFINITIONS})
+if (XGBOOST_MM_PREFETCH_PRESENT)
+  target_compile_definitions(objxgboost
+    PRIVATE
+    -DXGBOOST_MM_PREFETCH_PRESENT=1)
```
Comment (hcho3) [inline src/CMakeLists.txt]: What is the benefit of removing `build_config.h.in` and `configure_file()` ?

Comment (trivialfis) [reply src/CMakeLists.txt]: @hcho3 I prefer config file too.  One benefit of using something like `build_config.h` is to define macros that are visible to users, making them part of the interface.  If we were to employ config file instead of definitions, we should use it consistently for all defined macros and generate that file to build tree instead of source tree.  But that can not be done without first deprecating makefiles since make doesn't have a concept of `build tree vs source tree`.

So, using definitions for all macros is a temporary but consistent solution.

**Comment by trivialfis** [comment]
Ping @CodingCat

**Comment by trivialfis** [comment]
Thanks for all the help!
